### PR TITLE
Laboratory work #3, Veselov Sergei - 20FPL3

### PR DESCRIPTION
**Pull request by Azzruge**

### CODE REVIEW

**Comment by WhiteJaeger** [comment]
It is time to push some changes for the 3rd lab.

**Comment by WhiteJaeger** [comment]
It is time to push some changes for the 3rd lab.

File: lab_3/main.py [hunk 1]
```diff
@@ -20,8 +20,31 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+    split_regex = re.compile(r'[.|!|?|…]')
+    text = filter(lambda t: t, [t.strip() for t in split_regex.split(text)])
+    letters_result = []
+    shitty_symbols = ['`', '~', '!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '_', '-', '+',
+                        '=', '{', '[', ']', '}', '|', '\\', ':', ';', '"', "'", '<', ',', '>',
+                        '.', '?', '/', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0']
+    deutsch_buchstabe = {'ä': 'ae', 'ö': 'oe', 'ü': 'ue', 'ß': 'ss'}
+    for sentence in text:
+        sentence = sentence.lower()
+        for symbols in shitty_symbols:
```
Comment (WhiteJaeger) [inline lab_3/main.py:35]: `shitty` is not academic, please change.

File: lab_3/main.py [hunk 2]
```diff
@@ -20,8 +20,31 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+            sentence = sentence.replace(symbols, '')
+        tokens = sentence.split()
+        for word in tokens:
+            for letter, trans in deutsch_buchstabe.items():
+                word.replace(letter, trans)
+        letters = []
+        for token in tokens:
+            letters.append(tuple(['_'] + list(token) + ['_']))
+        if letters == []:
```
Comment (WhiteJaeger) [inline lab_3/main.py:44]: Can be done without explicit comparison.

File: lab_3/main.py [hunk 3]
```diff
@@ -116,26 +184,47 @@ def extract_n_grams(self, encoded_corpus: tuple) -> int:
...
+                if self.size >= len(token):
+                    tokens.append(token)
+                else:
+                    index = 0
+                    while index <= (len(token) - self.size):
+                        grams.append(tuple(token[index:index+self.size]))
+                        index += 1
+                    tokens.append(tuple(grams))
+            result.append(tuple(tokens))
+            self.n_grams = tuple(result)
```
Comment (WhiteJaeger) [inline lab_3/main.py:203]: Are you sure about it being inside the loop?

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_3/main.py (hunk 2 above)
Comment (Azzruge) [reply lab_3/main.py:44]: Should it looks like that 
if not letters
Did I understand it right?

File: lab_3/main.py [hunk 4]
```diff
@@ -234,11 +372,33 @@ def open(self, file_name: str) -> int:
...
         """
-        pass
+        if not isinstance(file_name, str):
+            return 1
+        with open(file_name, 'r', encoding="UTF-8") as file:
+            language_profile = json.load(file)
+        self.language = language_profile['name']
+        self.n_words = language_profile['n_words']
+        freq = language_profile['freq']
+        size = ''
```
Comment (WhiteJaeger) [inline lab_3/main.py:382]: Why do you instantiate it as a string if later it is compared with `len`?

File: lab_3/main.py [hunk 5]
```diff
@@ -213,16 +324,43 @@ def get_top_k_n_grams(self, k: int, trie_level: int) -> tuple:
...
+            if isinstance(corpus, tuple):
+                for element in corpus:
+                    string += self.storage.get_letter_by_id(element)
+            else:
+                return 1
+            freq[string] = freq_value
+        file_w = {'freq': freq, 'n_words': self.n_words, 'name': self.language}
+        with open(name, 'w', encoding='UTF-8') as file:
+            json_string = json.dumps(file_w)
+            file.write(json_string)
```
Comment (WhiteJaeger) [inline lab_3/main.py:362]: There is a `.dump` method in the `json` package that combines these two lines.

File: lab_3/main.py [hunk 6]
```diff
@@ -213,16 +324,43 @@ def get_top_k_n_grams(self, k: int, trie_level: int) -> tuple:
...
         :param name: name of the json file with .json format
         :return: 0 if profile saves, 1 if any errors occurred
         """
-        pass
+        if not isinstance(name, str):
+            return 1
+        freq_tuple = {}
+        freq = {}
+        for trie in self.tries:
+            freq_tuple |= trie.n_gram_frequencies
```
Comment (WhiteJaeger) [inline lab_3/main.py:350]: What does it do?

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_3/main.py (hunk 6 above)
Comment (Azzruge) [reply lab_3/main.py:350]: unites all n_gram_frequencies in one dict for simplifying the next loop

**Pull request closed without merging**